File: android/src/main/kotlin/com/pubmatic/sdk/flutter_openwrap_sdk/POBAdClient.kt
package com.pubmatic.sdk.flutter_openwrap_sdk

import com.pubmatic.sdk.openwrap.core.POBBid
import com.pubmatic.sdk.openwrap.core.POBImpression
import com.pubmatic.sdk.openwrap.core.POBRequest
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.Result

/**
 * Base class for all ad formats.
 */
abstract class POBAdClient(val adId: Int, protected val channel: MethodChannel) {

  protected var request: POBRequest? = null
  protected var impression: POBImpression? = null

  /**
   * Calls the respective OpenWrap SDK's API.
   *
   * @param names received via methodChannel
   * @param call to fetch the required arguments received from dart side
   * @param result to transfer the result for native side
   */
  open fun methodCall(names: List<String>, call: MethodCall, result: Result) {
    when (names[1]) {
      "setRequest" -> {
        setRequestProperties(call, request)
        result.success(null)
      }

      "setImpression" -> {
        setImpressionProperties(call, impression)
        result.success(null)
      }

      else -> result.notImplemented()
    }
  }

  /**
   * Sets the properties received from map to [POBRequest] of respective ad format.
   *
   * @param call [MethodCall] helps to fetch the argument came in method channel.
   * @param request [POBRequest] allocated to ads
   */
  private fun setRequestProperties(call: MethodCall, request: POBRequest?) {
    call.argument<Boolean>("debug")?.let {
      request?.enableDebugState(it)
    }

    call.argument<Int>("networkTimeout")?.let {
      request?.networkTimeout = it
    }

    call.argument<Boolean>("returnAllBidStatus")?.let {
      request?.enableReturnAllBidStatus(it)
    }

    // Deprecated in SDK version 4.5.0
    request?.versionId = call.argument<Int>("versionId")

    call.argument<Boolean>("testMode")?.let {
      request?.enableTestMode(it)
    }

    request?.adServerUrl = call.argument<String>("adServerUrl")
  }

  /**
   * Sets the properties received from map to [POBImpression] of respective ad format.
   *
   * @param call [MethodCall] helps to fetch the argument came in method channel.
   * @param impression [POBImpression] allocated to ads
   */
  private fun setImpressionProperties(call: MethodCall, impression: POBImpression?) {
    call.argument<Int>("adPosition")?.let {
      impression?.adPosition =
        POBRequest.AdPosition.values().getOrElse(it) { POBRequest.AdPosition.UNKNOWN }
    }

    impression?.testCreativeId = call.argument<String>("testCreativeId")

    impression?.setCustomParam(call.argument<Map<String, List<String>>>("customParams"))

    call.argument<String>("gpid")?.let { impression?.setGpid(it) }
  }

  /**
   * Convert the instance of [POBBid] to [HashMap]
   *
   * @param bid instance of [POBBid]
   * @return map with bid properties
   */
  protected fun convertBidToMap(bid: POBBid?): Map<String, Any?> {
    val bidMap = HashMap<String, Any?>()

    bidMap["bidId"] = bid?.id
    bidMap["impressionId"] = bid?.impressionId
    bidMap["bundle"] = bid?.bundle
    bidMap["price"] = bid?.price
    bidMap["height"] = bid?.height
    bidMap["width"] = bid?.width
    bidMap["status"] = bid?.status
    bidMap["creativeId"] = bid?.creativeId
    bidMap["nurl"] = bid?.getnURL()
    bidMap["lurl"] = bid?.getlURL()
    bidMap["creative"] = bid?.creative
    bidMap["creativeType"] = bid?.creativeType
    bidMap["partnerName"] = bid?.partnerName
    bidMap["dealId"] = bid?.dealId
    bidMap["refreshInterval"] = bid?.refreshInterval
    bidMap["targetingInfo"] = bid?.targetingInfo
    bidMap["rewardAmount"] = bid?.firstReward?.amount
    bidMap["rewardCurrencyType"] = bid?.firstReward?.currencyType

    return bidMap
  }

  /**
   * Removes the instance entry form [POBAdInstanceManager]'s map.
   */
  protected fun destroy() {
    POBAdInstanceManager.unRegisterAd(this)
  }
}

File: ios/Classes/POBAdClient.swift
import UIKit
import OpenWrapSDK
import Flutter

class POBAdClient: NSObject {
    private(set) var adId: Int
    var request: POBRequest?
    init(adId: Int) {
        self.adId = adId
    }

    /**
     Creates and returns `Dictionary` of`POBBid`

     - Parameters:
     - bid: `POBBid` received from Openwrap SDK

     - Returns: Dictionary<String, Any> containing fields from `POBBid`
     */
    func convertBidToMap(bid: POBBid?) -> [String: Any] {
        var bidMap = [String: Any]()
        guard let bid else { return bidMap }

        bidMap["bidId"] = bid.bidId
        bidMap["impressionId"] = bid.impressionId
        bidMap["bundle"] = bid.bundle
        bidMap["price"] = bid.price.doubleValue
        bidMap["height"] = Int(bid.size.height)
        bidMap["width"] = Int(bid.size.width)
        bidMap["status"] = bid.status.intValue
        bidMap["creativeId"] = bid.creativeId
        bidMap["nurl"] = bid.nurl
        bidMap["lurl"] = bid.lurl
        bidMap["creative"] = bid.creativeTag
        bidMap["creativeType"] = bid.creativeType
        bidMap["partnerName"] = bid.partner
        bidMap["dealId"] = bid.dealId
        bidMap["refreshInterval"] = Int(bid.refreshInterval)
        bidMap["targetingInfo"] = bid.targetingInfo()
        bidMap["rewardAmount"] = bid.reward()?.amount.intValue
        bidMap["rewardCurrencyType"] = bid.reward()?.currencyType
        return bidMap
    }

    private func setRequestProperties(call: FlutterMethodCall, request: POBRequest?) {
        guard let values = call.arguments as? [String: Any] else { return }
        // `values` dictionary has dynamic types. Hence explicit type casting is required to set properties of `POBRequest`
        // which have default value. From flutter plugin, it is ensured that these properties will always receive some value, either custom or default one.
        // Also, to follow coding guidelines, `if let` is preferred over forced type casting.
        if let debug = values["debug"] as? Bool {
            request?.debug = debug
        }
        if let networkTimeout = values["networkTimeout"] as? Int {
            request?.networkTimeout = TimeInterval(networkTimeout)
        }
        if let returnAllBidStatus = values["returnAllBidStatus"] as? Bool {
            request?.returnAllBidStatus = returnAllBidStatus
        }
        if let testModeEnabled = values["testMode"] as? Bool {
            request?.testModeEnabled = testModeEnabled
        }

        // These properties can have either custom value or default value as `null`
        // versionId API is deprecated in v4.5.0 OWSDK and will be removed in future SDK version.
        request?.versionId = values["versionId"] as? NSNumber
        request?.adServerURL = values["adServerUrl"] as? String
    }

    private func setImpressionProperties(call: FlutterMethodCall, impression: POBImpression?) {
        guard let values = call.arguments as? [String: Any] else { return }
        // `values` dictionary has dynamic types. Hence explicit type casting is required to set properties of `POBImpression`
        // which have default value. From flutter plugin, it is ensured that these properties will always receive some value, either custom or default one.
        // Also, to follow coding guidelines, `if let` is preferred over forced type casting.
        if let adPositionValue = values["adPosition"] as? Int {
            if adPositionValue < 2 {
                impression?.adPosition = POBAdPosition(rawValue: adPositionValue) ?? .unKnown
            } else {
                impression?.adPosition = POBAdPosition(rawValue: adPositionValue + 1) ?? .unKnown
            }
        }
        // These properties can have either custom value or default value as `null`
        impression?.testCreativeId = values["testCreativeId"] as? String
        impression?.customParams = values["customParams"] as? [String: Any]
        // Set gpid
        impression?.gpid = values["gpid"] as? String
    }

    func invokeMethod(names: [String], call: FlutterMethodCall, result: FlutterResult) {
        switch names[1] {
        case "setRequest":
            setRequestProperties(call: call, request: request)
            result(nil)
        case "setImpression":
            setImpressionProperties(call: call, impression: request?.impressions.first)
            result(nil)
        case "destroy":
            POBAdInstanceManager.shared.unregisterAd(adClient: self)
            result(nil)
        default:
            result(FlutterMethodNotImplemented)
        }
    }
}
